Using git `--work-tree` option blows up with ghooks

I know this is a weird issue, but I've just encountered that using `--work-tree` (as in `git --work-tree some/folder commit`) causes ghooks to blow up with the following:



```js
$ git --work-tree ./foo commit -m 'an example of --work-tree crashing'
module.js:339
    throw err;
    ^

Error: Cannot find module '<snip>/ghooks-test/foo/package'
    at Function.Module._resolveFilename (module.js:337:15)
    at Function.Module._load (module.js:287:25)
    at Module.require (module.js:366:17)
    at require (module.js:385:17)
    at commandFor (<snip>/ghooks-test/node_modules/ghooks/lib/runner.js:17:13)
    at run (<snip>/ghooks-test/node_modules/ghooks/lib/runner.js:8:17)
    at <snip>/ghooks-test/.git/hooks/pre-commit:7:20
    at Object.<anonymous> (<snip>/ghooks-test/.git/hooks/pre-commit:8:3)
    at Module._compile (module.js:435:26)
    at Object.Module._extensions..js (module.js:442:10)
```

I think it's a case of simply try/catching and just cracefully failing if the package.json cannot be found. I'm happy to work on a PR if you're happy to accept one? 

FYI if you're thinking "what in the world are you using --work-tree for?!" - well, I'm trying to have a gh-pages branch which tracks my `site` directory (rather than repo root) - using https://github.com/lukekarrys/git-directory-deploy

The problem with `--no-verify` is that while `pre-commit` and `commit-msg` hooks are ignored, `prepare-commit-msg` is not - and so even with `--no-verify`, failing hard means users still cannot commit. To demonstrate:

```js
$ git --work-tree ./foo commit -m 'an example of --work-tree crashing'
module.js:339
    throw err;
    ^

Error: Cannot find module '<snip>/ghooks-test/foo/package'
...
    at <snip>/ghooks-test/.git/hooks/pre-commit:7:20 <---- NOTE: pre-commit
...

$ git --work-tree ./foo commit -m 'an example of --work-tree crashing' --no-verify
module.js:339
    throw err;
    ^

Error: Cannot find module '<snip>/ghooks-test/foo/package'
...
    at <snip>/ghooks-test/.git/hooks/prepare-commit-msg:7:20 <---- NOTE: prepare-commit-msg
...
```

We could, I suppose, hard fail on some, while soft fail on others? `prepare-commit-msg` soft failing isn't the end of the world, because you just end up with an empty commit message (but hard failing means you can't commit).

--- a/.jshintrc
+++ b/.jshintrc
@@ -4,8 +4,8 @@
   "camelcase"     : true,
   "maxparams"     : 3,
   "maxdepth"      : 3,
-  "maxstatements" : 5,
-  "maxcomplexity" : 2,
+  "maxstatements" : 6,
+  "maxcomplexity" : 3,
   "maxlen"        : 80,
 
   "asi"           : false,

--- a/lib/install.js
+++ b/lib/install.js
@@ -3,9 +3,7 @@ module.exports = installHooks;
 var fs = require('fs');
 var cwd = process.cwd();
 var resolve = require('path').resolve;
-var projectRoot = cwd.replace(/node_modules(?!.*node_modules(\/|\\|$)).*/, '');
-projectRoot = resolve(projectRoot);
-var hooksDir = resolve(projectRoot, '.git/hooks');
+var findup = require('findup');
 var template = require('./hook.template');
 
 var hooks = [
@@ -28,16 +26,22 @@ var hooks = [
   'post-rewrite',
 ];
 
-function installHooks () {
-  if (isGitProject()) {
-    ensureHooksDirExists();
-    hooks.forEach(install);
+function findGitRoot (directory) {
+  try {
+    return findup.sync(directory || cwd, '.git');
+  } catch(e) {
+    return null;
   }
-  else { warnAboutGit(); }
 }
 
-function isGitProject () {
-  return fs.existsSync(resolve(projectRoot, '.git'));
+function installHooks (directory) {
+  var gitRoot = findGitRoot(directory);
+  if (gitRoot) {
+    var hooksDir = resolve(gitRoot, '.git/hooks');
+    ensureHooksDirExists(hooksDir);
+    hooks.forEach(install.bind(null, hooksDir));
+  }
+  else { warnAboutGit(); }
 }
 
 function warnAboutGit () { console.warn(
@@ -47,8 +51,8 @@ function warnAboutGit () { console.warn(
   'Please ignore this message if you are not using ghooks directly.'
 );}
 
-function install(hook) {
-  var file = hookFile(hook);
+function install(hooksDir, hook) {
+  var file = hookFile(hooksDir, hook);
   if (needsBackup(file)) { backup(file); }
   fs.writeFileSync(file, template.content);
   fs.chmodSync(file, '755');
@@ -66,10 +70,10 @@ function backup (file) {
   fs.renameSync(file, file + '.bkp');
 }
 
-function hookFile (name) {
+function hookFile (hooksDir, name) {
   return resolve(hooksDir, name);
 }
 
-function ensureHooksDirExists () {
+function ensureHooksDirExists (hooksDir) {
   if (!fs.existsSync(hooksDir)) { fs.mkdirSync(hooksDir); }
 }

--- a/lib/runner.js
+++ b/lib/runner.js
@@ -2,6 +2,8 @@ var clone = require('lodash.clone');
 var managePath = require('manage-path');
 var spawn = require('spawn-command');
 var resolve = require('path').resolve;
+var findup = require('findup');
+var fs = require('fs');
 
 module.exports = function run (dirname, filename, env) {
   var restorePath;
@@ -14,10 +16,10 @@ function hook (dirname, filename) {
 }
 
 function commandFor (hook) {
-  var pkg = require(resolve(process.cwd(), 'package'));
+  var pkgFile = findup.sync(process.cwd(), 'package.json');
+  var pkg = JSON.parse(fs.readFileSync(resolve(pkgFile, 'package.json')));
   if (pkg.config && pkg.config.ghooks && pkg.config.ghooks[hook]) {
     var command = pkg.config.ghooks[hook];
-
     // replace any instance of $1 or $2 etc. to that item as an process.argv
     return command.replace(/\$(\d)/g, function(match, number) {
       return process.argv[number];

--- a/package.json
+++ b/package.json
@@ -26,6 +26,7 @@
     "install": "node ./bin/install"
   },
   "dependencies": {
+    "findup": "^0.1.5",
     "lodash.clone": "3.0.3",
     "manage-path": "2.0.0",
     "spawn-command": "^0.0.2"

--- a/test/runner.test.js
+++ b/test/runner.test.js
@@ -1,5 +1,6 @@
 var path = require('path');
 var getPathVar = require('manage-path/dist/get-path-var');
+var findup = require('findup');
 require('./setup')();
 
 describe('runner', function () {
@@ -7,6 +8,11 @@ describe('runner', function () {
     var on = this.spawnOn = stub();
     this.spawn = spy(function () { return { on: on }; });
     this.run = proxyquire('../lib/runner', { 'spawn-command': this.spawn });
+    spy(findup, 'sync');
+  });
+
+  afterEach(function () {
+    findup.sync.restore();
   });
 
   beforeEach(setupPackageJsonWith({ config: {
@@ -17,6 +23,12 @@ describe('runner', function () {
       'post-merge': 'echo $PATH'
     }
   }}));
+
+  it('looks for `package.json` using findup, starting with process.cwd()', function () {
+    this.run('/foo/bar/', '/pre-commit');
+    expect(findup.sync)
+      .to.have.been.calledWith(process.cwd(), 'package.json');
+  });
 
   it('executes the command specified on the ghooks config', function () {
     this.run(process.cwd(), '/pre-commit');

